packages/server/src/workers/download.ts

```diff
@@ -253,28 +268,54 @@ export async function execDownloadJob<T extends Resource = Resource>(job: Job<Do
     // Note that while the Fetch Standard requires the property to always be a WHATWG ReadableStream, in node-fetch it is a Node.js Readable stream.
     await getBinaryStorage().writeBinary(binary, contentDisposition, contentType, response.body as Readable);
 
-    const updated = JSON.parse(JSON.stringify(resource).replace(url, `Binary/${binary.id}`)) as Resource;
-    (updated.meta as Meta).author = { reference: 'system' };
-    await systemRepo.updateResource(updated);
+    // re-fetch resource so we are mutating as recent a copy as possible
+    // (there may have been other mutations applied while we were writing the
+    // object into storage)
+    resource = await systemRepo.readResource<T>(resourceType, id);
+
+    const attachments = getAttachments(resource);
+    const patches = attachments
+      .filter(attachment => attachment.value.url === url)
+      .map(value => ({
+        op: 'replace' as const,
+        path: `${pathToPointer(value.path)}/url`,
+        value: `Binary/${binary.id}`,
+      }));
+    await systemRepo.patchResource(
+      resourceType,
+      id,
+      [
+        ...patches,
+        { op: 'replace', path: '/meta/author', value: { reference: 'system' } },
+      ],
+    );
+
     log.info('Downloaded content successfully');
   } catch (ex: any) {
     log.info('Download exception: ' + ex, ex);
     throw ex;
   }
 }
 
-function getAttachments(resource: Resource): Attachment[] {
-  const attachments: Attachment[] = [];
+// Translate basic FHIRPath into JSON Patch pointer
+export function pathToPointer(path: string): string {
+  return (path.startsWith('.') ? path : `.${path}`)
+    .replaceAll('.', '/')
+    .replaceAll(/\[(\d+)]/g, (m, g1) => `/${g1}`) // convert `[0]` => `/0`
+}
```

Comment:
This feels like something that should already have a more robust implementation somewhere, though I haven't found it yet. I'd appreciate references to existing work I can reuse.

Some alternative approaches:
- Teach the crawler to natively emit paths as JSON pointers
- Apply the FHIRPath directly instead of going through a JSON pointer at all. It looks like We could evaluate a path via (`evalFhirPath`)[https://github.com/medplum/medplum/blob/main/packages/core/src/fhirpath/parse.ts#L245], but it's not quite so obvious how to use this to patch the resource - we seem to only support JSON Patch in the API today.

Reply:
I've shifted this to be co-located with the crawler that emits these paths, and added some test coverage. @mattwiller, you mentioned that there may be more places where we have very similar code? My search didn't turn up similar implementations but I might be using poor search terms. If you have any pointers for anything you would like to see consolidated please let me know!

Commit here: https://github.com/medplum/medplum/pull/7781/commits/51e80fd0f46ba9f0824f7b24e4d6275d6b8148ab